File: app/src/main/java/com/group_project/chatapplication/ContactShowActivity.java
package com.group_project.chatapplication;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.Manifest;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.provider.ContactsContract;
import android.widget.Toast;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.database.FirebaseDatabase;

import java.util.ArrayList;
import java.util.HashMap;

public class ContactShowActivity extends AppCompatActivity {

    RecyclerView recycler_view;
    ArrayList<Model_Contact> arrayList = new ArrayList<Model_Contact>();
    Adapter_Contact adapter;
//    FirebaseDatabase database;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_contact_show);


        recycler_view = findViewById(R.id.recycler_view);

//        database = FirebaseDatabase.getInstance();

        checkPermission();

    }


    private void checkPermission() {
        if (ContextCompat.checkSelfPermission(ContactShowActivity.this, Manifest.permission.READ_CONTACTS) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(ContactShowActivity.this, new String[]{
                    Manifest.permission.READ_CONTACTS}, 100);
        } else {
            getContactList();
        }
    }

    private void getContactList() {
        Uri uri = ContactsContract.Contacts.CONTENT_URI;
        String sort = ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME + "ASC";
        Cursor cursor = getContentResolver().query(uri, null, null, null, sort);

        HashMap map = new HashMap();

        if (cursor.getCount() > 0) {
            while (cursor.moveToNext()) {
                String id = cursor.getString(cursor.getColumnIndex(ContactsContract.Contacts._ID));
                String name = cursor.getString(cursor.getColumnIndex(ContactsContract.Contacts.DISPLAY_NAME));
                Uri uriPhone = ContactsContract.CommonDataKinds.Phone.CONTENT_URI;
                String selection = ContactsContract.CommonDataKinds.Phone.CONTACT_ID + "=?";
                Cursor phoneCursor = getContentResolver().query(uriPhone, null, selection, new String[]{id}, null);

                if (phoneCursor.moveToNext()) {
                    String number = phoneCursor.getString(phoneCursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER));
                    Model_Contact model_contact = new Model_Contact();
                    model_contact.setContact_name(name);
                    model_contact.setContact_number(number);
                    arrayList.add(model_contact);

                    map.put(
                            phoneCursor.getString(phoneCursor.getColumnIndex((ContactsContract.CommonDataKinds.Phone.NUMBER))),
                            phoneCursor.getString(phoneCursor.getColumnIndex((ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME )))
                    );

                    phoneCursor.close();
                }
            }
            cursor.close();
        }
        recycler_view.setLayoutManager(new LinearLayoutManager(this));
        adapter = new Adapter_Contact(this, arrayList);
        recycler_view.setAdapter(adapter);

/*
        database.getReference().updateChildren(map).addOnSuccessListener(new OnSuccessListener() {
            @Override
            public void onSuccess(Object o) {
                Toast.makeText(getApplicationContext(), "Success!", Toast.LENGTH_SHORT).show();
            }
        }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception e) {
                Toast.makeText(getApplicationContext(), "Fail!", Toast.LENGTH_SHORT).show();
            }
        });
*/

    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == 100 && grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            getContactList();
        } else {
            Toast.makeText(getApplicationContext(), "Permission Denied.!", Toast.LENGTH_SHORT).show();
            checkPermission();
        }
    }

}

File: app/src/main/java/com/group_project/chatapplication/Contact_Show_Activity.kt
package com.group_project.chatapplication

import android.content.Context
import android.os.Bundle
import android.provider.ContactsContract
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase

class Contact_Show_Activity : AppCompatActivity() {
    lateinit var database: DatabaseReference
    lateinit var auth: FirebaseAuth
    lateinit var name: String
    lateinit var number: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_contact_show)

        val contact_list = findViewById<RecyclerView>(R.id.contact_list)
        contact_list.layoutManager = LinearLayoutManager(this)
        val contactList: MutableList<ContactDTO> = ArrayList()
        val map: HashMap<String, String> = HashMap<String, String>()

        auth = FirebaseAuth.getInstance()
        val login_phone = auth.currentUser?.phoneNumber.toString()
        database = FirebaseDatabase.getInstance().getReference("Contacts").child(login_phone)

        val contacts = contentResolver.query(
            ContactsContract.CommonDataKinds.Phone.CONTENT_URI,
            null,
            null,
            null,
            null
        )

        if (contacts != null) {
            while (contacts.moveToNext()) {
                name =
                    contacts.getString(contacts.getColumnIndex(ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME))
                number =
                    contacts.getString(contacts.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER))
                val obj = ContactDTO()
                obj.name = name
                obj.phone_number = number

                contactList.add(obj)

                map.put("contact_name", name)
                map.put("contact_number", number)
                database.push().setValue(map)
            }

            database.push().setValue(map)
                .addOnSuccessListener {
                    Toast.makeText(this, "Upload Successfully!", Toast.LENGTH_SHORT).show()
                }.addOnFailureListener {
                    Toast.makeText(this, "Not Upload!", Toast.LENGTH_SHORT).show()
                }

            contact_list.adapter = ContactAdapter(contactList, this)
            contacts.close()
        }

    }

    class ContactAdapter(items: MutableList<ContactDTO>, ctx: Context) :
        RecyclerView.Adapter<ContactAdapter.ViewHolder>() {

        private var list = items
        private var context = ctx

        override fun getItemCount(): Int {
            return list.size
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            holder.name.text = list[position].name
            holder.number.text = list[position].phone_number
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return ViewHolder(
                LayoutInflater.from(context).inflate(R.layout.item_contact, parent, false)
            )
        }

        class ViewHolder(v: View) : RecyclerView.ViewHolder(v) {
            val name: TextView = v.findViewById(R.id.tv_name)!!
            val number: TextView = v.findViewById(R.id.tv_number)!!

        }

    }

}